### ISSUE TITLE
FilePath stat cache cleaning should happen lazily, to avoid unnecessary stat() calls

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4130]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4130|
|Type|enhancement|
|Created|2009-11-24 00:47:10Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-changed-4130-3|

When calling, for example, `FilePath.remove()`, we have to re-stat the file to get correct behavior, otherwise the file still appears to exist.

However, maybe we don't care about what the file looks like any more.  Maybe we're never going to call `stat()` again, before the `FilePath` gets garbage collected.  If that is the case, then calling it again would be a waste.

There should be a method, perhaps `FilePath.changed()` [as implemented here](http://trac.calendarserver.org/browser/CalendarServer/trunk/lib-patches/Twisted/twisted.python.filepath.patch), for example, which just tells a `FilePath` to forget about its cache _without_ immediately getting a new cache.

<details><summary>Searchable metadata</summary>

```
trac-id__4130 4130
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_filepath_changed_4130_3 branches/filepath-changed-4130-3
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1259023630000000 1259023630000000
changetime__1274778933000000 1274778933000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

See also the unresolved debate on [#3161](https://github.com/twisted/twisted/issues/3161).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [28131]) Branching to 'filepath-changed-4130'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

Perhaps the simple implementation here will stimulate further discussion.

I do have to say that in the process of inspecting the current caching semantics I was _really_ disappointed.  I am tempted to say that we should kick the current caching logic to the curb, deprecate FilePath, and implement a new 2-layer solution: a completely naive no-caching-at-all layer, and a layer that has precise caching: for example, if you call isdir() three times, you should really get exactly one stat call regardless of the return value of that call.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @glyph|
|-|-|

  1. Please add
    1. A unit test which fails when `FilePath.moveTo` does not end with `self.changed()`
    1. A docstring to `FilePathTestCase.test_copyTo`, the only test which fails without the `self.changed()` in `FilePath.remove` _or_ a new test which covers that behavior less accidentally.
    1. `@ivar` for `statinfo` (I am particularly curious about the `None`/`0` deal).
    1. `@since` to `FilePath.changed`
    1. A news file
  1. Please update the copyright date on both modified files
  1. There is one pyflakes warning in `test_paths.py` (unrelated to your changes), if you care to fix it.

One other idle thought I had is whether `changed` might collide with any other method or attribute we might want to add to `FilePath` which relates to the actual file itself, rather than to the `FilePath` instance.  However, I can't quite nail down a specific idea of what it might collide with, and anyway any name might collide with something, so that's a silly thing to worry about.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [28996]) Respond to code review, point 1: unit test that fails when FilePath.moveTo does not end with self.changed().

Go beyond that a bit, because really moveTo also needs to do destination.changed(); add a test for that, and also some tests that will fail in case the implementation of EXDEV handling were ever to change its interaction with caching.

Refs [#4130](https://github.com/twisted/twisted/issues/4130)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_4):
> A docstring to `FilePathTestCase.test_copyTo`, the only test which fails without the `self.changed()` in `FilePath.remove` _or_ a new test which covers that behavior less accidentally.

I did hopefully improve the comprehensibility of the tests a bit, but this isn't quite true: `test_remove` also fails in the expected way.  (I renamed it to `test_removeDirectory` since that's really the case that it's testing directly.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

(In [28998]) Branching to 'filepath-changed-4130-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [29000]) Review feedback: @since annotation for new method.

(plus an unrelated docstring typo fix)

refs [#4130](https://github.com/twisted/twisted/issues/4130)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [29001]) @ivar for statinfo, as requested by code review.  It's documented, but you still don't want to use it.

refs [#4130](https://github.com/twisted/twisted/issues/4130)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [29003]) Yeah, statinfo = 0 is the wrong thing for changed() to do.

refs [#4130](https://github.com/twisted/twisted/issues/4130)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-changed-4130-2)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [29040]) Branching to 'filepath-changed-4130-3'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Making things a bit easier for the reviewer, here are some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-changed-4130-3) which shouldn't have noise from buildbot configuration and known failures.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set owner to @glyph|
|-|-|

OK nevermind.  Those now-less-noisy build results indicate failures on Windows which I now have to investigate, bleh.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph removed owner|
|-|-|

[Once more,](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-changed-4130-3) with _feeling_.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Okay, there was still a Windows + Python 2.4 failure due to the way that I wrote `test_changed`.  [This time for sure.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-changed-4130-3)

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|TimAllen set owner to @glyph|
|-|-|

Looks good to me, except that the documentation for the "statinfo" attribute is kind of confusing: generally "," has a higher precedence than ";" in English, and I was trying to read everything after the ";" up to the end of the sentence as a single logical unit.

I'd recommend wording something like this:

> The currently cached status information about the file on disk. This attribute is C{None} if the file is in an indeterminate state (either this L{FilePath} has not yet called C{stat()} or L{FilePath.changed} indicated that new information is required), 0 if C{stat()} was called and returned an error (i.e. the path did not exist when C{stat()} was called), or a stat result object that describes the status of the path at a particular point in time. Trust me when I tell you that you do not want to use this attribute; instead, use the methods on L{FilePath} which give you information about it.

With that change, +1 for merging.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [29046]) Better grammar for statinfo @ivar, as suggested by review.  refs [#4130](https://github.com/twisted/twisted/issues/4130)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

(In [29047]) Thanks to shadow-reviewer therve, change test docstrings to be affirmative (like I always tell everybody else to do).  refs [#4130](https://github.com/twisted/twisted/issues/4130).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set status to `closed`|
|-|-|

(In [29048]) Merge filepath-changed-4130-3: lazily refresh filesystem status

Author: glyph

Reviewer: exarkun, therve, screwtape

Fixes: [#4130](https://github.com/twisted/twisted/issues/4130)

FilePath now calls os.stat() only when new status information is required,
rather than immediately when anything changes.  For some applications this may
result in fewer stat() calls.  Additionally, FilePath has a new method,
'changed', which applications may use to indicate that the FilePath may have
been changed on disk and therefore the next status information request must
fetch a new stat result.  This is useful if external systems, such as C
libraries, may have changed files that Twisted applications are referencing via
a FilePath.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
